Add Subscribable and use it in subpackages

jnicklas opened the pull request:
Depends on #231 

This adds a `Subscribable` class, and uses it for `on` and `Channel`. It also adapts the other packages to use this new interface.

## Motivation

We noticed that in effection v2, working with subscriptions directly is somewhat cumbersome. The normal case is that the subscriptions will be "drained" through either `forEach` or `first` or similar. Since `forEach` uses internal iteration, the scope of the iterator can be constrained to the task that the `forEach` runs into. However, creating an iterator requires a task. If we create the iterator and then drain it, we need to pass in a task. If we do this all in one step we don't need to pass in this task. This makes the API much cleaner and easier to understand.

## Approach

We add a `Subscribable` interface and a `createSubscribable` function which creates subscribables. The hypothesis is that manually creating subscribables is rare, and that it should usually only be done for primitive structures (e.g. Channel). Therefore the user will get a convenient interface without having to think too much about the intrecacies of this system.

Here's an example of what the current API looks like:

``` typescript
task.spawn(on<[string]>(task, childProcess.stdout, 'data').forEach(([data]) => function*() {
  // ...
}));
```

Note the `task` passed to `on`. Here is the same code with the new API:

``` typescript
task.spawn(on<[string]>(childProcess.stdout, 'data').forEach(([data]) => {
  // ...
}));
```

### Alternate Designs

The status quo is an alternative, where we double down on working with `Subscription` rather than `Subscribable`, but for reasons explained above this feels inconvenient.

### Possible Drawbacks or Risks

Subscriptions are naturally race condition proof, whereas subscribables are not. Some care needs to be taken that no events/messages are lost before subscribing. However, this risk is lessened in effection v2, since task spawning no longer is an operation.

### TODOs and Open Questions

- [ ] Should we still have `Subscription` at all? That is should we have some object with higher level operations like `map` and `filter` for an actual subscription object, rather than a subscribable object? At the moment we feel that having such an object is probably *not* necessary, but more real-world experience is necessary.

jnicklas committed:
Add Subscribable and use it in subpackages

This adds a `Subscribable` class, and uses it for `on` and `Channel`. It
also adapts the other packages to use this new interface.

taras commented:
@jnicklas would it be possible to add before/after code example to the description?

jnicklas committed:
Don't require operation in forEach

It's quite common for the iteration callback to `forEach` to not
actually have to be an operation, since it doesn't need to yield. This
way we can provide a void returning function instead.

jnicklas commented:
@taras I added an example to the description!

taras commented:
@jnicklas thank you. That helps a lot!
